syncer/error.go

```diff
@@ -77,55 +81,159 @@ func originError(err error) error {
+	invalidConnF := func(tctx *tcontext.Context, err error, ddls []string, index int, conn *DBConn) error {
+		// must ensure only the last statement executed failed with the `invalid connection` error
+		if len(ddls) == 0 || index != len(ddls)-1 || errors.Cause(err) != mysql.ErrInvalidConn {
+			return err // return the original error
+		}
 
-	handle := func() {
-		tctx.L().Warn("ignore special error for DDL", zap.String("DDL", ddl2), log.ShortError(err))
-		err2 := conn.resetConn(tctx) // also reset the `invalid connection` for later use.
+		ddl2 := ddls[index]
+		stmt, err2 := parser2.ParseOneStmt(ddl2, "", "")
 		if err2 != nil {
-			tctx.L().Warn("reset connection failed", log.ShortError(err2))
+			return err // return the original error
+		}
+
+		handle := func() {
+			tctx.L().Warn("ignore special error for DDL", zap.String("DDL", ddl2), log.ShortError(err))
+			err2 := conn.resetConn(tctx) // also reset the `invalid connection` for later use.
+			if err2 != nil {
+				tctx.L().Warn("reset connection failed", log.ShortError(err2))
+			}
 		}
+
+		switch v := stmt.(type) {
+		case *ast.AlterTableStmt:
+			// ddls should be split with only one spec
+			if len(v.Specs) > 1 {
+				return err
+			} else if v.Specs[0].Tp == ast.AlterTableAddConstraint {
+				// only take effect on `ADD INDEX`, no UNIQUE KEY and FOREIGN KEY
+				// UNIQUE KEY may affect correctness, FOREIGN KEY should be filtered.
+				// ref https://github.com/pingcap/tidb/blob/3cdea0dfdf28197ee65545debce8c99e6d2945e3/ddl/ddl_api.go#L1929-L1948.
+				switch v.Specs[0].Constraint.Tp {
+				case ast.ConstraintKey, ast.ConstraintIndex:
+					handle()
+					return nil // ignore the error
+				}
+			}
+		case *ast.CreateIndexStmt:
+			handle()
+			return nil // ignore the error
+		}
+		return err
 	}
 
-	switch v := stmt.(type) {
-	case *ast.AlterTableStmt:
-		// ddls should be split with only one spec
-		if len(v.Specs) > 1 {
-			return err
-		} else if v.Specs[0].Tp == ast.AlterTableAddConstraint {
-			// only take effect on `ADD INDEX`, no UNIQUE KEY and FOREIGN KEY
-			// UNIQUE KEY may affect correctness, FOREIGN KEY should be filtered.
-			// ref https://github.com/pingcap/tidb/blob/3cdea0dfdf28197ee65545debce8c99e6d2945e3/ddl/ddl_api.go#L1929-L1948.
-			switch v.Specs[0].Constraint.Tp {
-			case ast.ConstraintKey, ast.ConstraintIndex:
-				handle()
-				return nil // ignore the error
+	// for DROP COLUMN with its single-column index, try drop index first then drop column
+	dropColumnF := func(tctx *tcontext.Context, originErr error, ddls []string, index int, conn *DBConn) error {
+		mysqlErr, ok := originError(originErr).(*mysql.MySQLError)
+		if !ok {
+			return originErr
+		}
+		// different version of TiDB has different error message, try to cover most versions
+		if !(mysqlErr.Number == errno.ErrUnsupportedDDLOperation &&
+			strings.Contains(mysqlErr.Message, "drop column") &&
+			strings.Contains(mysqlErr.Message, "with index")) {
+			return originErr
+		}
```

Comment:
how about defining a function to judge whether the error is `drop column with index`?

Reply:
changed in https://github.com/pingcap/dm/pull/801/commits/31f6b5a204a67509673f5d1b1d9e19365bae5ad3